### PR TITLE
[clang][tidy] Ensure rewriter has the correct CWD

### PR DESCRIPTION
**Pull request by jansvoboda11**
This patch replaces use of the deprecated `FileEntry::getName()` with `FileEntryRef::getName()`. This means the code now uses the path that was used to register file entry in `SourceManager` instead of the absolute path that happened to be used in the last call to `FileManager::getFile()` some place else.

This caused some test failures due to the fact that some paths can be relative and thus rely on the VFS CWD. The CWD can change for each TU, so when we run `clang-tidy` on a compilation database and try to perform all the replacements at the end, relative paths won't resolve the same. This patch takes care to reinstate the correct CWD and make the path reported by `FileEntryRef` absolute before passing it to `llvm::writeToOutput()`.

### CODE REVIEW

**Review by PiotrZSL — APPROVED** [review]
Clang part looks fine.
For a clang-tidy part, is there a way to test this part ? What changes because we use now a absolute build directory.
I'm missing some tests for that part.

**Comment by jansvoboda11** [comment]
> Clang part looks fine. For a clang-tidy part, is there a way to test this part ? What changes because we use now a absolute build directory. I'm missing some tests for that part.

So the clang-tidy part is necessary to keep the `clang-tidy/infrastructure/clang-tidy-run-with-database.cpp` test passing. It uses compilation database that has entries like this:

```json
{
  "directory": "<absolute-dir>/b",
  "command": "clang++ -o test.o ../b/c.cpp",
  "file": "<absolute-dir>/b/c.cpp"
}
```

I think this already provides good enough code coverage for this PR. What do you think?

What I'd like to get clarified in this PR is which directory to use as the CWD.
Is using `ClangTidyContext::getCurrentBuildDirectory()` okay, or do we need to look somewhere else?

**Comment by jansvoboda11** [comment]
Ping @PiotrZSL.

File: clang-tools-extra/clang-tidy/ClangTidy.cpp [hunk 1]
```diff
@@ -227,6 +227,14 @@ class ErrorReporter {
           llvm::errs() << "Can't apply replacements for file " << File << "\n";
         }
       }
+
+      auto BuildDir = Context.getCurrentBuildDirectory();
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/ClangTidy.cpp:231]: I think you may need to do what is done in handleErrors method.
Simply remember old value, change new value to build directory, and after call to overwriteChangedFiles restore it.
By default getCurrentWorkingDirectory should return actually current working directory.

Comment (jansvoboda11) [reply clang-tools-extra/clang-tidy/ClangTidy.cpp]: You're right, the `clang-tidy/infrastructure/clang-tidy-run-with-database.cpp` test kept passing just by chance. I added extra test case to prove that we need to keep closer track for each file what was the CWD when we processed it.

File: clang-tools-extra/clang-tidy/ClangTidy.cpp [hunk 2]
```diff
@@ -189,9 +191,12 @@ class ErrorReporter {
 
   void finish() {
     if (TotalFixes > 0) {
-      Rewriter Rewrite(SourceMgr, LangOpts);
+      bool AnyNotWritten = false;
       for (const auto &FileAndReplacements : FileReplacements) {
+        Rewriter Rewrite(SourceMgr, LangOpts);
         StringRef File = FileAndReplacements.first();
+        Files.getVirtualFileSystem().setCurrentWorkingDirectory(
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/ClangTidy.cpp:198]: maybe would be nice to restore CurrentWorkingDirectory after this function come to the end.

**Review by PiotrZSL — APPROVED** [review]
LGTM

**Pull request merged**